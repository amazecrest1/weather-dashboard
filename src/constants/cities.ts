import { City, ParameterConfig } from '../types/weather';

// Predefined cities with coordinates
export const CITIES: City[] = [
  {
    id: 'new-york',
    name: 'New York',
    country: 'United States',
    latitude: 40.7128,
    longitude: -74.0060,
  },
  {
    id: 'london',
    name: 'London',
    country: 'United Kingdom',
    latitude: 51.5074,
    longitude: -0.1278,
  },
  {
    id: 'tokyo',
    name: 'Tokyo',
    country: 'Japan',
    latitude: 35.6762,
    longitude: 139.6503,
  },
  {
    id: 'sydney',
    name: 'Sydney',
    country: 'Australia',
    latitude: -33.8688,
    longitude: 151.2093,
  },
  {
    id: 'paris',
    name: 'Paris',
    country: 'France',
    latitude: 48.8566,
    longitude: 2.3522,
  },
  {
    id: 'dubai',
    name: 'Dubai',
    country: 'United Arab Emirates',
    latitude: 25.2048,
    longitude: 55.2708,
  },
];

// Default city (New York)
export const DEFAULT_CITY = CITIES[0];

// Hourly parameter configurations for detailed insights
export const HOURLY_PARAMETERS: ParameterConfig[] = [
  {
    key: 'temperature_2m',
    label: 'Temperature',
    unit: '°C',
    color: '#ef4444',
    yAxisId: 'temperature',
  },
  {
    key: 'relative_humidity_2m',
    label: 'Relative Humidity',
    unit: '%',
    color: '#3b82f6',
    yAxisId: 'humidity',
  },
  {
    key: 'apparent_temperature',
    label: 'Apparent Temperature',
    unit: '°C',
    color: '#f59e0b',
    yAxisId: 'temperature',
  },
  {
    key: 'precipitation',
    label: 'Precipitation',
    unit: 'mm',
    color: '#06b6d4',
    yAxisId: 'precipitation',
  },
  {
    key: 'pressure_msl',
    label: 'Sea Level Pressure',
    unit: 'hPa',
    color: '#8b5cf6',
    yAxisId: 'pressure',
  },
  {
    key: 'wind_speed_10m',
    label: 'Wind Speed (10m)',
    unit: 'km/h',
    color: '#10b981',
    yAxisId: 'wind',
  },
];

// Chart colors
export const CHART_COLORS = {
  temperature: {
    max: '#ef4444',
    min: '#3b82f6',
    mean: '#f59e0b',
  },
  precipitation: '#06b6d4',
  windSpeed: '#10b981',
  humidity: '#8b5cf6',
  pressure: '#f97316',
};

// API endpoints
export const API_BASE_URL = 'https://archive-api.open-meteo.com/v1/archive';

// Date range limits
export const MAX_DATE_RANGE_DAYS = 90; // 3 months

// Default date range (last 7 days)
export const getDefaultDateRange = () => {
  const end = new Date();
  const start = new Date();
  start.setDate(end.getDate() - 7);
  
  return {
    start: start.toISOString().split('T')[0],
    end: end.toISOString().split('T')[0],
  };
}; 