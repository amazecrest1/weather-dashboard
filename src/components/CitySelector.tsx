import React, { useState, useEffect, useRef } from 'react';
import { City } from '../types/weather';
import { CITIES } from '../constants/cities';
import { debouncedSearchCities, saveCustomCity, getCustomCities, removeCustomCity } from '../utils/geocoding';

interface CitySelectorProps {
  selectedCity: City;
  onCityChange: (city: City) => void;
  className?: string;
}

const ALL_CITIES_OPTION: City = {
  id: 'all-cities',
  name: 'All Cities Selected',
  country: '',
  latitude: 0,
  longitude: 0,
};

const CitySelector: React.FC<CitySelectorProps> = ({ 
  selectedCity, 
  onCityChange, 
  className = '' 
}) => {
  const [searchQuery, setSearchQuery] = useState('');
  const [searchResults, setSearchResults] = useState<City[]>([]);
  const [customCities, setCustomCities] = useState<City[]>([]);
  const [isOpen, setIsOpen] = useState(false);
  const [selectedIndex, setSelectedIndex] = useState(-1);
  
  const searchRef = useRef<HTMLInputElement>(null);
  const dropdownRef = useRef<HTMLDivElement>(null);

  // Load custom cities on mount
  useEffect(() => {
    setCustomCities(getCustomCities());
  }, []);

  // Handle search input changes
  useEffect(() => {
    if (searchQuery.trim().length >= 2) {
      debouncedSearchCities(searchQuery, (cities) => {
        setSearchResults(cities);
      });
    } else {
      setSearchResults([]);
    }
  }, [searchQuery]);

  // All available cities (predefined + custom)
  const allCities = [ALL_CITIES_OPTION, ...CITIES, ...customCities];

  // Filtered cities based on search query
  const filteredCities = searchQuery.length < 2 
    ? allCities 
    : allCities.filter(city => 
        city.name.toLowerCase().includes(searchQuery.toLowerCase()) ||
        city.country.toLowerCase().includes(searchQuery.toLowerCase())
      );

  // Combined suggestions (filtered existing cities + search results)
  const suggestions = [
    ...filteredCities,
    ...searchResults.filter(result => 
      !allCities.some(city => 
        Math.abs(city.latitude - result.latitude) < 0.01 && 
        Math.abs(city.longitude - result.longitude) < 0.01
      )
    )
  ];

  const handleInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setSearchQuery(e.target.value);
    setSelectedIndex(-1);
  };

  const handleCitySelect = (city: City) => {
    // Save custom city if it's not already saved
    if (city.isCustom && !customCities.some(c => c.id === city.id)) {
      saveCustomCity(city);
      setCustomCities(prev => [...prev, city]);
    }
    
    onCityChange(city);
    setSearchQuery('');
    setIsOpen(false);
    setSelectedIndex(-1);
  };

  const handleKeyDown = (e: React.KeyboardEvent) => {
    if (!isOpen) return;

    switch (e.key) {
      case 'ArrowDown':
        e.preventDefault();
        setSelectedIndex(prev => 
          prev < suggestions.length - 1 ? prev + 1 : prev
        );
        break;
      case 'ArrowUp':
        e.preventDefault();
        setSelectedIndex(prev => prev > 0 ? prev - 1 : prev);
        break;
      case 'Enter':
        e.preventDefault();
        if (selectedIndex >= 0 && selectedIndex < suggestions.length) {
          handleCitySelect(suggestions[selectedIndex]);
        }
        break;
      case 'Escape':
        setIsOpen(false);
        setSelectedIndex(-1);
        setSearchQuery('');
        break;
    }
  };

  const handleRemoveCustomCity = (cityId: string, e: React.MouseEvent) => {
    e.stopPropagation();
    removeCustomCity(cityId);
    setCustomCities(prev => prev.filter(c => c.id !== cityId));
    
    // If the removed city was selected, switch to default
    if (selectedCity.id === cityId) {
      onCityChange(ALL_CITIES_OPTION);
    }
  };

  const handleDropdownClick = () => {
    setIsOpen(!isOpen);
    if (!isOpen) {
      setTimeout(() => {
        searchRef.current?.focus();
      }, 100);
    }
  };

  const handleDropdownBlur = () => {
    setTimeout(() => {
      setIsOpen(false);
      setSelectedIndex(-1);
      setSearchQuery('');
    }, 200);
  };

  // Update display value
  const displayValue = selectedCity.id === 'all-cities'
    ? 'All Cities Selected'
    : selectedCity.name + (selectedCity.state ? `, ${selectedCity.state}` : '') + `, ${selectedCity.country}`;

  return (
    <div className={`relative flex flex-col ${className}`} style={{ zIndex: 1000 }}>
      <div className="relative">
        <button
          onClick={handleDropdownClick}
          className={`w-full px-6 py-3 border border-gray-200 rounded-xl bg-white text-gray-900 focus:outline-none focus:ring-2 focus:ring-blue-500 focus:border-blue-500 text-left flex items-center justify-between shadow-lg hover:border-gray-300 transition-colors font-medium`}
          style={{ minHeight: 56 }}
        >
          <span className="truncate">{displayValue}</span>
          <span className="flex items-center justify-center ml-3">
            <svg className="w-6 h-6 text-gray-400" fill="none" stroke="currentColor" strokeWidth="2" viewBox="0 0 24 24">
              <path strokeLinecap="round" strokeLinejoin="round" d="M19 9l-7 7-7-7" />
            </svg>
          </span>
        </button>
      </div>

      {/* Dropdown */}
      {isOpen && (
        <div
          ref={dropdownRef}
          className="absolute z-[9999] top-full mt-2 w-full bg-white border border-gray-200 rounded-xl shadow-lg"
          style={{
            boxShadow: '0 8px 32px 0 rgba(31, 38, 135, 0.10)',
            border: '1px solid #e5e7eb',
            marginTop: '8px',
            minWidth: 260,
          }}
        >
          {/* Search input */}
          <div className="p-3 border-b border-gray-100 bg-gray-50 rounded-t-xl">
            <input
              ref={searchRef}
              type="text"
              value={searchQuery}
              onChange={handleInputChange}
              onKeyDown={handleKeyDown}
              onBlur={handleDropdownBlur}
              placeholder="Search for a city..."
              className="w-full px-3 py-2 border border-gray-200 rounded-md focus:outline-none focus:ring-2 focus:ring-blue-500 focus:border-blue-500 bg-white"
            />
          </div>
          
          {/* Results */}
          <div className="max-h-60 overflow-y-auto">
            {suggestions.length === 0 ? (
              <div className="px-3 py-3 text-gray-400 text-sm text-center">
                {searchQuery.length < 2 ? 'Type to search for cities' : 'No cities found'}
              </div>
            ) : (
              <div>
                {suggestions.map((city, index) => (
                  <div
                    key={city.id}
                    onClick={() => handleCitySelect(city)}
                    className={`px-4 py-3 cursor-pointer flex items-center justify-between transition-colors border-b border-gray-100 last:border-b-0 rounded-lg mx-2 my-1 ${
                      index === selectedIndex
                        ? 'bg-blue-50 text-blue-700'
                        : 'hover:bg-gray-100'
                    } ${city.id === 'all-cities' ? 'font-semibold text-gray-700 bg-gray-100 hover:bg-blue-50' : ''}`}
                  >
                    <div className="flex flex-col flex-1 min-w-0">
                      <span className="font-medium text-gray-900 truncate">
                        {city.name}
                        {city.state && <span className="text-gray-500">, {city.state}</span>}
                      </span>
                      {city.id !== 'all-cities' && (
                        <span className="text-sm text-gray-500 truncate">{city.country}</span>
                      )}
                    </div>
                    <div className="flex items-center gap-2 flex-shrink-0">
                      {city.isCustom && (
                        <span className="text-xs bg-blue-100 text-blue-700 px-2 py-0.5 rounded-full font-medium">
                          Custom
                        </span>
                      )}
                      {city.isCustom && customCities.some(c => c.id === city.id) && (
                        <button
                          onClick={(e) => handleRemoveCustomCity(city.id, e)}
                          className="text-gray-400 hover:text-red-500 p-1 rounded-full hover:bg-red-50 transition-colors"
                          title="Remove custom city"
                        >
                          <svg className="w-4 h-4" fill="none" stroke="currentColor" viewBox="0 0 24 24">
                            <path strokeLinecap="round" strokeLinejoin="round" strokeWidth="2" d="M6 18L18 6M6 6l12 12" />
                          </svg>
                        </button>
                      )}
                    </div>
                  </div>
                ))}
              </div>
            )}
          </div>
        </div>
      )}
    </div>
  );
};

export default CitySelector; 