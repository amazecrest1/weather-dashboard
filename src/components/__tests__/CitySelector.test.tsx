import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import CitySelector from '../CitySelector';
import { CITIES } from '../../constants/cities';

describe('CitySelector', () => {
  const mockOnCityChange = jest.fn();
  const defaultProps = {
    selectedCity: CITIES[0],
    onCityChange: mockOnCityChange,
    customCities: [],
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders the city dropdown button with selected city', () => {
    render(<CitySelector {...defaultProps} />);
    const button = screen.getByRole('button');
    expect(button).toBeInTheDocument();
    expect(button).toHaveTextContent('New York');
    expect(button).toHaveTextContent('United States');
  });

  it('shows dropdown options when button is clicked', () => {
    render(<CitySelector {...defaultProps} />);
    fireEvent.click(screen.getByRole('button'));
    expect(screen.getByText('All Cities Selected')).toBeInTheDocument();
    expect(screen.getByText('New York')).toBeInTheDocument();
    expect(screen.getByText('United States')).toBeInTheDocument();
  });

  it('calls onCityChange when a city is selected', () => {
    render(<CitySelector {...defaultProps} />);
    fireEvent.click(screen.getByRole('button'));
    // Find the city option by label, then click it
    const londonOption = screen.getByText('London');
    expect(londonOption).toBeInTheDocument();
    // Optionally, check its country is rendered as a sibling
    const country = londonOption.parentElement?.querySelector('.text-sm');
    expect(country).toHaveTextContent('United Kingdom');
    fireEvent.click(londonOption);
    expect(mockOnCityChange).toHaveBeenCalledWith(
      expect.objectContaining({ name: 'London', country: 'United Kingdom' })
    );
  });

  it('calls onCityChange when All Cities Selected is chosen', () => {
    render(<CitySelector {...defaultProps} />);
    fireEvent.click(screen.getByRole('button'));
    const allCitiesOption = screen.getByText('All Cities Selected');
    expect(allCitiesOption).toBeInTheDocument();
    fireEvent.click(allCitiesOption);
    expect(mockOnCityChange).toHaveBeenCalledWith(
      expect.objectContaining({ id: 'all-cities' })
    );
  });

  it('applies custom className', () => {
    render(<CitySelector {...defaultProps} className="custom-class" />);
    // The outermost wrapper should have the class
    const wrapper = screen.getByRole('button').closest('div.relative.flex.flex-col');
    expect(wrapper).toHaveClass('custom-class');
  });

  it('matches snapshot with default state', () => {
    const { container } = render(<CitySelector {...defaultProps} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot with dropdown open', () => {
    const { container } = render(<CitySelector {...defaultProps} />);
    fireEvent.click(screen.getByRole('button'));
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot with different selected city', () => {
    const { container } = render(
      <CitySelector 
        selectedCity={CITIES[1]} 
        onCityChange={mockOnCityChange} 
      />
    );
    expect(container.firstChild).toMatchSnapshot();
  });
}); 