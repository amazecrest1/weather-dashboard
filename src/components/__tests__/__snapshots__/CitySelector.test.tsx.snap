// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CitySelector matches snapshot with default state 1`] = `
<div
  class="relative flex flex-col "
  style="z-index: 1000;"
>
  <div
    class="relative"
  >
    <button
      class="w-full px-3 sm:px-4 md:px-6 py-3 border border-gray-200 dark:border-gray-600 rounded-xl bg-white dark:bg-gray-700 text-gray-900 dark:text-emerald-200 bg-gradient-to-r from-emerald-500 to-teal-500 bg-clip-text text-transparent focus:outline-none focus:ring-2 focus:ring-emerald-500 dark:focus:ring-emerald-400 focus:border-emerald-500 dark:focus:border-emerald-400 text-left flex items-center justify-between shadow-lg dark:shadow-gray-900/50 hover:border-gray-300 dark:hover:border-gray-500 transition-all duration-200 font-medium text-xs sm:text-sm md:text-base backdrop-blur-sm"
      style="min-height: 56px;"
    >
      <span
        class="truncate flex-1 min-w-0 text-left"
      >
        New York, United States
      </span>
      <span
        class="flex items-center justify-center ml-2 sm:ml-3 flex-shrink-0"
      >
        <svg
          class="w-4 h-4 sm:w-5 sm:h-5 md:w-6 md:h-6 text-gray-400"
          fill="none"
          stroke="currentColor"
          stroke-width="2"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 9l-7 7-7-7"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </span>
    </button>
  </div>
</div>
`;

exports[`CitySelector matches snapshot with different selected city 1`] = `
<div
  class="relative flex flex-col "
  style="z-index: 1000;"
>
  <div
    class="relative"
  >
    <button
      class="w-full px-3 sm:px-4 md:px-6 py-3 border border-gray-200 dark:border-gray-600 rounded-xl bg-white dark:bg-gray-700 text-gray-900 dark:text-emerald-200 bg-gradient-to-r from-emerald-500 to-teal-500 bg-clip-text text-transparent focus:outline-none focus:ring-2 focus:ring-emerald-500 dark:focus:ring-emerald-400 focus:border-emerald-500 dark:focus:border-emerald-400 text-left flex items-center justify-between shadow-lg dark:shadow-gray-900/50 hover:border-gray-300 dark:hover:border-gray-500 transition-all duration-200 font-medium text-xs sm:text-sm md:text-base backdrop-blur-sm"
      style="min-height: 56px;"
    >
      <span
        class="truncate flex-1 min-w-0 text-left"
      >
        London, United Kingdom
      </span>
      <span
        class="flex items-center justify-center ml-2 sm:ml-3 flex-shrink-0"
      >
        <svg
          class="w-4 h-4 sm:w-5 sm:h-5 md:w-6 md:h-6 text-gray-400"
          fill="none"
          stroke="currentColor"
          stroke-width="2"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 9l-7 7-7-7"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </span>
    </button>
  </div>
</div>
`;

exports[`CitySelector matches snapshot with dropdown open 1`] = `
<div
  class="relative flex flex-col "
  style="z-index: 1000;"
>
  <div
    class="relative"
  >
    <button
      class="w-full px-3 sm:px-4 md:px-6 py-3 border border-gray-200 dark:border-gray-600 rounded-xl bg-white dark:bg-gray-700 text-gray-900 dark:text-emerald-200 bg-gradient-to-r from-emerald-500 to-teal-500 bg-clip-text text-transparent focus:outline-none focus:ring-2 focus:ring-emerald-500 dark:focus:ring-emerald-400 focus:border-emerald-500 dark:focus:border-emerald-400 text-left flex items-center justify-between shadow-lg dark:shadow-gray-900/50 hover:border-gray-300 dark:hover:border-gray-500 transition-all duration-200 font-medium text-xs sm:text-sm md:text-base backdrop-blur-sm"
      style="min-height: 56px;"
    >
      <span
        class="truncate flex-1 min-w-0 text-left"
      >
        New York, United States
      </span>
      <span
        class="flex items-center justify-center ml-2 sm:ml-3 flex-shrink-0"
      >
        <svg
          class="w-4 h-4 sm:w-5 sm:h-5 md:w-6 md:h-6 text-gray-400"
          fill="none"
          stroke="currentColor"
          stroke-width="2"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 9l-7 7-7-7"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </span>
    </button>
  </div>
</div>
`;
